### PR TITLE
[Blocked Security Review]Fixes #1144 Sharing of session amongst tabs and sharing of Authenticated Documents

### LINKED ISSUE #1144: Print button is unresponsive for Online PDF bank statements
### Description: 
Print button is unresponsive for Online PDF bank statements

### Steps to Reproduce 
  1. Clean profile 1.9.3
  2. Open online PDF bank statements
  3. Tap on the Action button at the bottom in the iOS device
  4. Tap on "Print" button 

**Note**: Please note this issue is happening only for online PDF bank statements.

**Actual result:** 
The print button becomes unresponsive and it doesn't open the print preview page.

**Expected result:**
Tap on "Print" button should show a print preview page and allow the user to print the PDF

**Reproduces how often:** [Easily reproduced, Intermittent Issue]
Always

**Brave Version:** 
1.9.3 (19.05.22.15)

**Device details:** 
iPhone 8 - iOS 12.3.1

**Website problems only:**
- did you check with Brave Shields down? Yes - It did not resolve the issue
- did you check in Safari/Firefox (WkWebView-based browsers)? Yes - works fine in safari

### Additional Information 
cc: @brave/legacy_qa @jhreis
The issue is reported by the community user
https://community.brave.com/t/print-button-unresponsive-on-ipad/64083

### PR DESCRIPTION
**Pull request by Brandon-T**
**BUG FIXES:**

- Fixes: https://github.com/brave/brave-ios/issues/1144
- Fixed a background thread issue where "displayTitle" is being called on WKWebView on a non-main thread causing the PrintPreview to have undefined-behaviour.
- Fixed opening Authenticated documents in Private-Browsing mode when it opens in a new tab.
- Fixed opening Authenticated documents and attempting to share, or print them.
- Fixes tabs not having the same data store as other tabs (IE: Log-In to any website on one tab should log you in on another tab).
- Fixed destroying of data-store to only happen when ALL tabs are destroyed.
- Changed Temporary Storage to pull out the document from the WebPage instead of making server calls outside the webview's session (IE: No need for cookie injecting, etc).. Instead, we get the webview to give us the document via Javascript injection which passes the document to iOS for display/saving, etc..

**WHAT I TESTED:**
- Log into Email (gmail.com) in PrivateTab-A
   - Refresh PrivateTab-B
   - PrivateTab-B is also logged in.

- Log out of Email (gmail.com) in PrivateTab-A
   - Refresh PrivateTab-B
   - PrivateTab-B is logged out.

- Above tested in normal mode as well.

- Log into PrivateTab-A
   - Open Normal Tab
   - Normal tab is NOT logged in.
   - Switch back to private mode, all tabs are gone.

- Log into Normal Tab-A
   - Open PrivateTab
   - Private Tab is not logged in
   - Switch back to Normal Tab - Still logged in.

- Log into banking website in PrivateTab-A
   - View statements (It opens statements a PDF in PrivateTab-B [ authenticated document ])
   - Success (PDF opened).
   - Share PDF in PrivateTab-B (Works).

- Log into banking website in PrivateTab-A
   - Switched to Normal mode and back to PrivateMode.
   - Tabs killed, cookies and sessions are cleared successfully for all Private Tabs.

- Log into banking website in PrivateTab-A
   - Open second PrivateTab-B.
   - Kill both Tabs.
   - Cookies and sessions are cleared successfully for all Private Tabs.

- Log into banking website in PrivateTab-A
   - View statements as PDF (opens in a new PrivateTab-B).
   - Turn Internet OFF
   - Share PDF and Print (works without internet because no requests are made)
   - Kill tabs
   - Cookies and sessions are cleared successfully for all Private Tabs.

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-ios/issues) for my issue if one did not already exist.
- [x] My patch or PR title has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!` (or `No Bug: <message>` if no relevant ticket)
- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New files have MPL-2.0 license header.

## Reviewer Checklist:

- [x] PR is linked to an issue via [Zenhub](https://www.zenhub.com/extension).
- [ ] Issues are assigned to at least one epic.
- [x] Issues include necessary QA labels:
  - [x] `QA/(Yes|No)`
  - [ ] `release-notes/(include|exclude)`
  - [x] `bug` / `enhancement`
- [x] Necessary security reviews have taken place.
- [ ] Adequate test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable)

### CODE REVIEW

File: Client/Frontend/Browser/TabManager.swift [hunk 1]
```diff
@@ -528,7 +528,11 @@ class TabManager: NSObject {
         }
 
         if tab.isPrivate {
-            removeAllBrowsingDataForTab(tab)
+            // Only when ALL tabs are dead, we clean up.
+            // This is because other tabs share the same data-store.
+            if allTabs.filter({ $0.isPrivate }).count <= 1 {
```
Comment (danishjafri88) [inline Client/Frontend/Browser/TabManager.swift]: you can use `tabsForCurrentMode` instead of `allTabs.filter({ $0.isPrivate })`

File: Client/Frontend/Browser/TemporaryDocument.swift [hunk 2]
```diff
@@ -47,14 +44,52 @@ class TemporaryDocument: NSObject {
...
+            UIApplication.shared.isNetworkActivityIndicatorVisible = false
+        }
+        
+        // Store the blob/data in a local temporary file.
+        if let document = document, let data = document.data, !data.isEmpty {
+            let tempDirectory = URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent("TempDocs")
+            let url = tempDirectory.appendingPathComponent(filename)
+            
+            do {
+                try FileManager.default.createDirectory(at: tempDirectory, withIntermediateDirectories: true, attributes: nil)
```
Comment (danishjafri88) [inline Client/Frontend/Browser/TemporaryDocument.swift:72]: perhaps use try? if not catching any error?

**Review by danishjafri88 — CHANGES_REQUESTED** [review]

**Comment by danishjafri88** [comment]
Security review issue: https://github.com/brave/internal/issues/616

File: Client/Frontend/Browser/ResourceDownloadManager.swift [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    func scriptMessageHandlerName() -> String? {
+        return "resourceDownloadManager"
+    }
+    
+    func userContentController(_ userContentController: WKUserContentController, didReceiveScriptMessage message: WKScriptMessage) {
+        
+        do {
+            let response = try DownloadedResourceResponse.from(message: message)
+            tab?.temporaryDocument?.onDocumentDownloaded(document: response, error: nil)
+        } catch let error {
```
Comment (iccub) [inline Client/Frontend/Browser/ResourceDownloadManager.swift]: `let error` can be removed, `error` variable is available in catch closure for free

File: Client/Frontend/Browser/ResourceDownloadManager.swift [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+            tab?.temporaryDocument?.onDocumentDownloaded(document: nil, error: error)
+        }
+    }
+    
+    static func downloadResource(for tab: Tab, url: URL) {        
+        let token = UserScriptManager.securityToken.uuidString.replacingOccurrences(of: "-", with: "", options: .literal)
```
Comment (iccub) [inline Client/Frontend/Browser/ResourceDownloadManager.swift:57]: Maybe add a comment why we have to do this replacement?

File: Client/Frontend/Browser/ResourceDownloadManager.swift (hunk 3 above)
Comment (iccub) [inline Client/Frontend/Browser/ResourceDownloadManager.swift]: let error can be removed, error variable is available in catch closure for free

**Review by iccub — CHANGES_REQUESTED** [review]
looks good overall, small changes requested

File: Client/Frontend/Browser/ResourceDownloadManager.swift (hunk 3 above)
Comment (Brandon-T) [reply Client/Frontend/Browser/ResourceDownloadManager.swift]: Fixed.

File: Client/Frontend/Browser/TemporaryDocument.swift (hunk 2 above)
Comment (Brandon-T) [reply Client/Frontend/Browser/TemporaryDocument.swift:72]: I need the error to fall through so I don't repeat code in the catch block and so the "promise" doesn't get fulfilled with the download url.

File: Client/Frontend/Browser/TabManager.swift (hunk 1 above)
Comment (Brandon-T) [reply Client/Frontend/Browser/TabManager.swift]: Fixed.

**Comment by jumde** [comment]
There are too many edge cases with cookie handling in private and normal tabs, I think we should add more tests for the cases that we have tested with this PR to avoid regressions in future.

File: Client/Frontend/Browser/ResourceDownloadManager.swift [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    
+    static func from(message: WKScriptMessage) throws -> DownloadedResourceResponse? {
+        let data = try JSONSerialization.data(withJSONObject: message.body, options: .prettyPrinted)
+        return try JSONDecoder().decode(DownloadedResourceResponse.self, from: data)
+    }
+    
+    init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        self.statusCode = try container.decode(Int.self, forKey: .statusCode)
+        self.data = Data(base64Encoded: try container.decode(String.self, forKey: .base64Data))
```
Comment (jumde) [inline Client/Frontend/Browser/ResourceDownloadManager.swift:22]: minor-nit: we should probably add a maximum cap on how much data can be downloaded.

File: Client/Frontend/Browser/TemporaryDocument.swift (hunk 2 above)
Comment (danishjafri88) [reply Client/Frontend/Browser/TemporaryDocument.swift:72]: Oh yeah :)

**Comment by danishjafri88** [comment]
> There are too many edge cases with cookie handling in private and normal tabs, I think we should add more tests for the cases that we have tested with this PR to avoid regressions in future.

@jumde can you please help us getting together a list for the same. Its really difficult to figure out all possibilities :)

**Comment by jumde** [comment]
> > There are too many edge cases with cookie handling in private and normal tabs, I think we should add more tests for the cases that we have tested with this PR to avoid regressions in future.
> 
> @jumde can you please help us getting together a list for the same. Its really difficult to figure out all possibilities :)

@danishjafri88 - 

Yes! Let's use this comment to compile the list, feel free to add the cases you think i've missed:

1. Navigate to http://site-a.com/store.html in private tab
2. Open http://site-a.com in normal tab and confirm that no storage is available.

---- 

1. Navigate to http://site-a.com/store.html in normal tab
2. Open http://site-a.com in private tab and confirm no storage is available
3. Open http://site-a.com using long press and confirm that no storage is available.

---- 

1. Navigate to http://site-a.com/store.html in private tab
2. Close all tabs 
3. Confirm that no-storage is available for http://site-a.com in private tab

---- 

1. Navigate to http://site-a.com/store.html in private tab
2. Open http://site-a.com/ in a different private tab
3. Verify that storage is available for site-a in the second tab.

File: Client/Frontend/Browser/ResourceDownloadManager.swift (hunk 5 above)
Comment (Brandon-T) [reply Client/Frontend/Browser/ResourceDownloadManager.swift:22]: I can add that cap in Javascript but what would be the max file size? It will pull the "PDF" or image or whatever data from the WebKit's internal cache if available instead of "downloading" and it only does this when the user hits "share".

Also, what kind of tests did you have in mind? :)
I'll try to think of scenarios that I haven't tested already.
I've mostly been comparing various browsers' behaviours to our Brave-iOS.
Didn't realize how many problems arrises just trying to fix PDF :D lol

Comment (jumde) [reply Client/Frontend/Browser/ResourceDownloadManager.swift:22]: You can use the `Content-Length` header, if the request is from the cache you can check if the status code is 304.

File: Client/Frontend/UserContent/UserScripts/ResourceDownloader.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+            webkit.messageHandlers.resourceDownloadManager.postMessage(msg);
+        }
+    }
+                               
+    function downloadPageResource(link) {
+        var xhr = new XMLHttpRequest();
+        xhr.responseType = "arraybuffer";
+        xhr.onreadystatechange = function() {
+            if (this.readyState == XMLHttpRequest.DONE) {
+                if (this.status == 200) {
```
Comment (jumde) [inline Client/Frontend/UserContent/UserScripts/ResourceDownloader.js:19]: What about redirects here?

**Comment by Brandon-T** [comment]
> 1. Navigate to http://site-a.com/store.html in private tab
> 2. Open http://site-a.com in normal tab and confirm that no storage is available.
> 
> 1. Navigate to http://site-a.com/store.html in normal tab
> 2. Open http://site-a.com in private tab and confirm no storage is available
> 3. Open http://site-a.com using long press and confirm that no storage is available.
> 
> 1. Navigate to http://site-a.com/store.html in private tab
> 2. Close all tabs
> 3. Confirm that no-storage is available for http://site-a.com in private tab
> 
> 1. Navigate to http://site-a.com/store.html in private tab
> 2. Open http://site-a.com/ in a different private tab
> 3. Verify that storage is available for site-a in the second tab.


Hmm I tested these cases in my PR with gmail & safari dev tools.

Case 1 -> Opening a normal tab kills all private tabs and all their non-persistent stores along with them. Also tested in PR.

Case 2 -> Login gmail via Normal tab does NOT log you into a Private tab & vice-versa since they use different storage. Tested this in my PR.

Case 3 -> Killing the last private tab does the same as killing all private tabs and their non-persistent stores along with them.

Case 4 -> Verified in my PR via gmail log in. Login to tab-a automatically logs you in to tab-b.

File: Client/Frontend/UserContent/UserScripts/ResourceDownloader.js (hunk 6 above)
Comment (Brandon-T) [reply Client/Frontend/UserContent/UserScripts/ResourceDownloader.js:19]: It shouldn't matter because it only triggers the Javascript when you press "Share" and it only injects the Javascript when the page is completely loaded (IE: 200 or 404).
In such a case, you won't be able to share a 404 page anyway as the Mime-Type is not a doc type.

There is some code that was written before that checks: 
`if response.mimeType?.isKindOfHTML == false` before allowing you to "share" the page.
Basically, if the document was never loaded into a page with a 200, this Javascript won't ever be triggered because you won't be able to share in the first place.

In the case that the page is loaded and we attempt to get the document, I can't see it ever returning a redirect since the document is already downloaded. Also, if it does return a redirect for the weirdest possible reason, the JS would return you NULL as per the else statement.

File: Client/Frontend/Browser/ResourceDownloadManager.swift (hunk 5 above)
Comment (Brandon-T) [reply Client/Frontend/Browser/ResourceDownloadManager.swift:22]: The cache doesn't return a 304 ever (it's local). It returns a 0 if it fails. If that's the case, my code returns you null because there is no document available. In the case where a document is available to share, it returns a 200 always and the byte-data associated with it.

If the web-server returns a 304, the page would have loaded the document from the cache, this Javascript would pull it from the page returning a 200.

I checked Brave Desktop, there's no limit to what you can download and I haven't seen any limits in the old code before this PR so I'm not sure what the limit is.

**Review by iccub — APPROVED** [review]
Looks good, although this needs some serious testing from the QA team.

**Review by danishjafri88 — APPROVED** [review]
lgtm

**Comment by jumde** [comment]
> > 1. Navigate to http://site-a.com/store.html in private tab
> > 2. Open http://site-a.com in normal tab and confirm that no storage is available.
> > 3. Navigate to http://site-a.com/store.html in normal tab
> > 4. Open http://site-a.com in private tab and confirm no storage is available
> > 5. Open http://site-a.com using long press and confirm that no storage is available.
> > 6. Navigate to http://site-a.com/store.html in private tab
> > 7. Close all tabs
> > 8. Confirm that no-storage is available for http://site-a.com in private tab
> > 9. Navigate to http://site-a.com/store.html in private tab
> > 10. Open http://site-a.com/ in a different private tab
> > 11. Verify that storage is available for site-a in the second tab.
> 
> Hmm I tested these cases in my PR with gmail & safari dev tools.
> 
> Case 1 -> Opening a normal tab kills all private tabs and all their non-persistent stores along with them. Also tested in PR.
> 
> Case 2 -> Login gmail via Normal tab does NOT log you into a Private tab & vice-versa since they use different storage. Tested this in my PR.
> 
> Case 3 -> Killing the last private tab does the same as killing all private tabs and their non-persistent stores along with them.
> 
> Case 4 -> Verified in my PR via gmail log in. Login to tab-a automatically logs you in to tab-b.

I was thinking more of some automated tests, you can check the tests in `ClientTests` for reference. There are too many edge cases with storage, so just want to make sure we don't regress going forward.

**Comment by Brandon-T** [comment]
> I was thinking more of some automated tests, you can check the tests in `ClientTests` for reference. There are too many edge cases with storage, so just want to make sure we don't regress going forward.

Fixed. I've added some important tests that cleared up a lot of confusion and discovered some bugs along the way.

```Swift
class TabSessionTests: XCTestCase {
    /** Tests whether or not two private tabs are sharing the same data-store/cookie-store **/
    func testPrivateTabSessionSharing()

    /** Tests whether or not a private tab's data-store/cookie-store is truly non-persistent **/
    func testPrivateTabNonPersistence()

    /** Tests whether or not switching from private tab to normal tab destroys the data-store/cookie-store completely **/
    func testTabsPrivateToNormal()

    /** Tests whether or not switching from normal tab to private tab keeps the data-store/cookie-store separate **/
    func testTabsNormalToPrivate()
}
```

These seem to be the most important and have fixed any bugs related to their failures.
The last unit test has some code commented out because I couldn't solve why it was happening and I believe it to be a WebKit bug and not brave code related.

File: ClientTests/TabSessionTests.swift [hunk 7]
```diff
@@ -0,0 +1,527 @@
...
+        
+        destroyData { }
+        
+        tabManager = nil
+    }
+    
+    private func destroyData(_ completion: @escaping () -> Void) {
+        tabManager.removeAll()
+        tabManager.resetConfiguration()
+        tabManager.resetProcessPool()
```
Comment (danishjafri88) [inline ClientTests/TabSessionTests.swift]: We don't have to resetProcessPool, in fact we should avoid this as it can cause unwanted behaviour(crash - known) when working with parent-child tabs. The process pool is part of configurations so if you reset config the processPool is reset too.

Comment (danishjafri88) [reply ClientTests/TabSessionTests.swift]: Also calling `reset()` solves the need to call `resetConfiguration()`

File: ClientTests/TabSessionTests.swift [hunk 8]
```diff
@@ -0,0 +1,527 @@
...
+        let group = DispatchGroup()
+        
+        group.enter()
+        History.deleteAll({
+            group.leave()
+        })
+        
+        group.enter()
+        HTTPCookieStorage.shared.removeCookies(since: .distantPast)
+        WKWebsiteDataStore.default().httpCookieStore.getAllCookies({
```
Comment (danishjafri88) [inline ClientTests/TabSessionTests.swift]: You don't have to delete cookie explicitly here if you are calling `WKWebsiteDataStore.default().removeData(ofTypes: WKWebsiteDataStore.allWebsiteDataTypes()` below

File: ClientTests/TabSessionTests.swift [hunk 9]
```diff
@@ -0,0 +1,527 @@
...
+                group.leave()
+            })
+            self.tabManager.addNavigationDelegate(webViewNavigationAdapter)
+            
+            group.notify(queue: .main) {
+                // All requests finished loading.. time to check the cookies..
+                let group = DispatchGroup()
+                self.tabManager.allTabs.forEach({
+                    group.enter()
+                    $0.webView?.configuration.websiteDataStore.fetchDataRecords(ofTypes: WKWebsiteDataStore.allWebsiteDataTypes(), completionHandler: { records in
```
Comment (danishjafri88) [inline ClientTests/TabSessionTests.swift]: I am not sure, but will this not also check if the cookies are there or not. If so isn' the check below
`$0.webView?.configuration.websiteDataStore.httpCookieStore.getAllCookies(` redundant?

**Review by danishjafri88 — CHANGES_REQUESTED** [review]

File: ClientTests/TabSessionTests.swift (hunk 8 above)
Comment (Brandon-T) [reply ClientTests/TabSessionTests.swift]: > You don't have to delete cookie explicitly here if you are calling `WKWebsiteDataStore.default().removeData(ofTypes: WKWebsiteDataStore.allWebsiteDataTypes()` below

I don't know if it's a bug or not.. but if I don't, the tests are somehow keeping cookies from the test before it!

This is why I call it to make sure each test has a fair chance of working properly. So far this is the only fix that I've found that guarantees they are deleted no matter what.

File: ClientTests/TabSessionTests.swift (hunk 9 above)
Comment (Brandon-T) [reply ClientTests/TabSessionTests.swift]: Yes, but I'm not only checking for cookies.. I'm checking for disk cache, mem cache, etc.
The separate call for just cookies alone works too. I don't know if the websiteDataStore records and cookieStore are linked. That's why I did both just in case.

Are you sure they're linked or? I will write a quick test and see.

**EDIT:** You're right. A quick test shows that both are somehow linked. I know on iOS 11 they weren't for sure (at least my previous tests said they weren't).

File: ClientTests/TabSessionTests.swift (hunk 8 above)
Comment (danishjafri88) [reply ClientTests/TabSessionTests.swift]: Ah this bug, check `HttpCookieExtension.swift saveToDisk`. We might have to do `fetchDataRecords` here too? In either case, I understand now.

File: ClientTests/TabSessionTests.swift (hunk 9 above)
Comment (danishjafri88) [reply ClientTests/TabSessionTests.swift]: I think so, can you please verify :)

**Comment by danishjafri88** [comment]
@jumde tests have been added. Can you verify and update the security review status?

**Comment by iccub** [comment]
If you haven't done it already, this also needs to be rebased to latest development to pass CI

File: ClientTests/TabSessionTests.swift (hunk 9 above)
Comment (Brandon-T) [reply ClientTests/TabSessionTests.swift]: Fixed. You are correct.

File: ClientTests/TabSessionTests.swift (hunk 8 above)
Comment (Brandon-T) [reply ClientTests/TabSessionTests.swift]: I figured it out. It was actually because the non-persistent store was still ALIVE..

**Review by danishjafri88 — APPROVED** [review]
Solid work!!!

**Review by jumde — APPROVED** [review]
